examples/dnn_self_supervised_learning_ex.cpp

```diff
@@ -0,0 +1,412 @@
+                item.avg_green_second = item.avg_green_first;
+                item.avg_blue_second = item.avg_blue_first;
+            }
+            // read and discard the sizes if this was really a sized input layer.
+            if (version == "input_rgb_image_sized")
+            {
+                size_t nr, nc;
+                deserialize(nr, in);
+                deserialize(nc, in);
+            }
+        }
+
+        friend std::ostream& operator<<(std::ostream& out, const input_rgb_image_pair& item)
+        {
+            out << "input_rgb_image_pair(("
+                << item.avg_red_first<<","<<item.avg_green_first<<","<<item.avg_blue_first << "),("
+                << item.avg_red_second<<","<<item.avg_green_second<<","<<item.avg_blue_second << "))";
+            return out;
+        }
+
+        friend void to_xml(const input_rgb_image_pair& item, std::ostream& out)
+        {
+            out << "<input_rgb_image_pair r1='"<<item.avg_red_first<<"' g1='"<<item.avg_green_first<<"' b1='"<<item.avg_blue_first
+                << ",r2='"<<item.avg_red_second<<"' g2='"<<item.avg_green_second<<"' b2='"<<item.avg_blue_second<<"'/>";
+        }
+
+    private:
+        float avg_red_first, avg_red_second;
+        float avg_green_first, avg_green_second;
+        float avg_blue_first, avg_blue_second;
+    };
+}
+
+namespace dlib
+{
+    class loss_barlow_twins_
+    {
+    public:
+
+        loss_barlow_twins_() = default;
+
+        loss_barlow_twins_(float lambda) : lambda(lambda)
+        {
+            DLIB_CASSERT(lambda > 0);
+        }
+
+        template <
+            typename SUBNET
+        >
+        double compute_loss_value_and_gradient (
+            const tensor& input_tensor,
+            SUBNET& sub
+        ) const
+        {
+            const tensor& output_tensor = sub.get_output();
+            tensor& grad = sub.get_gradient_input();
+
+            DLIB_CASSERT(sub.sample_expansion_factor() == 2);
+            DLIB_CASSERT(input_tensor.num_samples() != 0);
+            DLIB_CASSERT(input_tensor.num_samples() % sub.sample_expansion_factor() == 0);
+            DLIB_CASSERT(input_tensor.num_samples() == grad.num_samples());
+            DLIB_CASSERT(input_tensor.num_samples() == output_tensor.num_samples());
+            DLIB_CASSERT(output_tensor.nr() == 1 && output_tensor.nc() == 1);
+            DLIB_CASSERT(grad.nr() == 1 && grad.nc() == 1);
+
+            const auto batch_size = output_tensor.num_samples() / 2;
+            const auto sample_size = output_tensor.k();
+            // std::cout << "batch size: " << batch_size << std::endl;
+            // std::cout << "sample size: " << sample_size << std::endl;
+
+            // some alias helpers to access the samples in the batch
+            alias_tensor split(batch_size, sample_size);
+            alias_tensor sample(1, sample_size);
+            auto za = split(output_tensor);
+            auto zb = split(output_tensor, batch_size);
```

Comment:
Is this what you want?  That is, offsetting the memory by `batch_size`?  Not some larger size?  Like this is going to overlap `za` isn't it?

Reply:
Ah! right, it has to be `batch_size` * `sample_size`... so that I jump to the second half of the batch...
It still crashes, but that was definitely wrong, thank you!